R/slice.R

```diff
@@ -165,6 +174,93 @@ vec_as_index <- function(i, n, names = NULL) {
   vec_assert(n, integer(), 1L)
   .Call(vctrs_as_index, i, n, names)
 }
+#' @rdname vec_as_index
+#' @export
+vec_as_position <- function(i, n, names = NULL) {
+  if (is.object(i)) {
```

Comment:
Is this too flexible? I'd want `vec_as_position(factor("a"))` to be an error. At least I think you should try coercing to character first.

Reply:
Either we restrict indices to be non-S3, or we coerce to int || chr. Factor indices come up because of stringsAsFactors, I got a request in tidyselect to make it work, and these are the same semantics that we discussed about over there.

I don't think the order matters because an S3 vector that is coercible to integer shouldn't be coercible to character. And for the rare case that it is, I think we should start with the most specific type integer before trying the least specific.

Reply:
The current behaviour gives exactly the wrong behaviour for factors. I don’t think any reasoning from tidyselect would apply here as this is a much lower level tool.

Reply:
What is the wrong behaviour for factors?

Reply:
`list(a = 1, b = 2)[[factor(“b”)]]`

Reply:
This is with base R rules for coercion. I think it makes sense to treat factors as character when indexing, which is what is implemented in this PR based on vctrs coercion rules.

Tidyselect may be high level, but selecting positions is a low level task. The behaviour should be consistent with vctrs (and so it should become an error if we disallow S3 indices in vctrs).

Reply:
Ok, I get it now.